File: port_CSharp/grun4py.cs
// ******* GRUN (Grammar Unit Test) for Python *******

using System;
using System.Text.RegularExpressions;
using System.Text;
using Antlr4.Runtime;

namespace grun4py
{
    internal static class Program
    {
        public static int Main(string[] args)
        {
            if (args.Length < 1)
            {
                Console.Error.WriteLine("Error: Please provide an input file path");
                return 1;
            }

            try
            {
                var filePath = args[0];
                var input = GetEncodedInputStreamByPythonComment(filePath);
                var lexer = new PythonLexer(input);
                var tokens = new CommonTokenStream((ITokenSource)lexer);
                var parser = new PythonParser(tokens);

                tokens.Fill(); // Test the lexer grammar
                foreach (IToken t in tokens.GetTokens())
                {
                    Console.WriteLine(GetTokenMetaDataWithTokenName(t));
                }

                parser.file_input(); // Test the parser grammar
                return parser.NumberOfSyntaxErrors;

            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                return 1; // Error occurred, returning non-zero exit code
            }
        }

        private static string GetTokenMetaDataWithTokenName(IToken token)
        {
            string tokenText = ReplaceSpecialCharacters(token.Text);
            string tokenName = token.Type == TokenConstants.EOF ? "EOF" : PythonLexer.DefaultVocabulary.GetDisplayName(token.Type);
            string channelText = token.Channel == TokenConstants.DefaultChannel ? 
                                 "" :
                                 $"channel={PythonLexer.channelNames[token.Channel]},";

            // Modified format: [@TokenIndex,StartIndex:StopIndex='Text',<TokenName>,channel=ChannelName,Line:Column]
            return $"[@{token.TokenIndex},{token.StartIndex}:{token.StopIndex}='{tokenText}',<{tokenName}>,{channelText}{token.Line}:{token.Column}]";
        }

        private static string ReplaceSpecialCharacters(string text)
        {
            return text.Replace("\n", @"\n")
                       .Replace("\r", @"\r")
                       .Replace("\t", @"\t")
                       .Replace("\f", @"\f");

        }

        public static ICharStream? GetEncodedInputStreamByPythonComment(string filePath)
        {
            string encodingName = "";
            var ws_commentPattern = new Regex(@"^[ \t\f]*(#.*)?$");

            try
            {
                using FileStream fs = new(filePath, FileMode.Open, FileAccess.Read); // read in binary mode
                using StreamReader reader = new(fs, Encoding.ASCII);
                for (int lineCount = 0; lineCount < 2; lineCount++)
                {
                    string? line = reader.ReadLine();
                    if (line == null)
                    {
                        break; // EOF reached
                    }

                    if (ws_commentPattern.IsMatch(line)) // WS? + COMMENT? found
                    {
                        encodingName = GetEncodingName(line);
                        if (encodingName != "") // encoding found
                        {
                            break;
                        }
                    }
                    else
                    {
                        break; // statement or backslash found (line is not empty, not whitespace(s), not comment)
                    }
                }
            }
            catch (Exception)
            {
                // Console.WriteLine($"An error occurred: {e.Message}");
            }

            const string DEFAULT_PYTHON_ENCODING = "utf-8"; // default encoding for Python source code
            if (encodingName == "")
            {
                encodingName = DEFAULT_PYTHON_ENCODING;
            }

            try // encoding test for ANTLR4
            {
                return CharStreams.fromPath(filePath, Encoding.GetEncoding(encodingName));
            }
            catch (Exception)
            {
                return CharStreams.fromPath(filePath, Encoding.GetEncoding(DEFAULT_PYTHON_ENCODING));
            }

        }

        public static string GetEncodingName(string commentText) // https://peps.python.org/pep-0263/#defining-the-encoding
        {
            var encodingCommentPattern = new Regex(@"^[ \t\f]*#.*?coding[:=][ \t]*([-_.a-zA-Z0-9]+)");
            var match = encodingCommentPattern.Match(commentText);
            if (match.Success)
            {
                string encodingName = match.Groups[1].Value;

                // normalize encoding name
                var encodingMap = new Dictionary<string, string>
                {
                    { "cp1252", "latin1" },
                    { "latin-1", "latin1" },
                    { "iso-8859-1", "latin1" }
                    // more encoding pairs
                };

                return encodingMap.TryGetValue(encodingName.ToLower(), out var normalizedEncodingName)
                    ? normalizedEncodingName
                    : encodingName;
            }
            return "";
        }
    }
}

File: port_JavaScript/grun4py.js
// ******* GRUN (Grammar Unit Test) for Python *******

import * as fs from 'fs';
import { CharStreams, CommonTokenStream, Token } from "antlr4";
import PythonLexer from "./PythonLexer.js";
import PythonParser from "./PythonParser.js";

class grun4py {
    static replaceSpecialCharacters(text) {
        return text.replace(/\n/g, "\\n")
                   .replace(/\r/g, "\\r")
                   .replace(/\t/g, "\\t")
                   .replace(/\f/g, "\\f");
    
    }
    
    static getTokenMetaDataWithTokenName(token) {
        const tokenText = this.replaceSpecialCharacters(token.text);
        const tokenName = token.type == Token.EOF ? "EOF" : PythonLexer.symbolicNames[token.type] ?? "";
        const channelText = token.channel == Token.DEFAULT_CHANNEL ?
            "" :
            `channel=${PythonLexer.channelNames[token.channel]},`;
    
        // original format: [@tokenIndex,start:stop='tokenText',<tokenType>,channel=channelNumber,line:column]
        // modified format: [@tokenIndex,start:stop='tokenText',<tokenName>,channel=channelName,line:column]
        return `[@${token.tokenIndex},${token.start}:${token.stop}='${tokenText}',<${tokenName}>,${channelText}${token.line}:${token.column}]`;
    }
    
    static async getEncodedInputStreamByPythonComment(filePath) {
        let encodingName = "";
        const ws_commentPattern = /^[ \t\f]*(#.*)?$/;
    
        try {
            const data = fs.readFileSync(filePath, { encoding: 'ascii' }); // read in binary mode
            const lines = data.split('\n');
    
            for (let lineCount = 0; lineCount < 2; lineCount++) { // only the first two lines
                let line = lines[lineCount];
                if (line === undefined) {
                    break; // EOF reached
                }
    
                line = line.replace('\r', '');
                if (ws_commentPattern.test(line)) { // WS* + COMMENT? found
                    encodingName = this.getEncodingName(line);
                    if (encodingName) {
                        break; // encoding found
                    }
                } else {
                    break; // statement or backslash found (line is not empty, not whitespace(s), not comment)
                }
            }
        } catch (error) {
            // console.error("An error occurred: " + error);
        }
    
        const DEFAULT_PYTHON_ENCODING = "utf-8"; // default encoding for Python source code 
        if (!encodingName) {
            encodingName = DEFAULT_PYTHON_ENCODING;
        }
    
        try { // encoding test for ANTLR4
            return CharStreams.fromPathSync(filePath, encodingName);
        } catch (error) {
            return CharStreams.fromPathSync(filePath, DEFAULT_PYTHON_ENCODING)
        }
    }
    
    static getEncodingName(commentText) { // https://peps.python.org/pep-0263/#defining-the-encoding
        const encodingCommentPattern = /^[ \t\f]*#.*?coding[:=][ \t]*([-_.a-zA-Z0-9]+)/;
        const match = commentText.match(encodingCommentPattern);
        if (match) {
            const encodingName = match[1];
    
            // normalize encoding name
            const encodingMap = {
                "cp1252": "latin1",
                "latin-1": "latin1",
                "iso-8859-1": "latin1"
                // more encoding pairs
            };
    
            return encodingMap[encodingName.toLowerCase()] || encodingName;
        }
        return "";
    }
    
    static async main(filePath) {
        if (process.argv.length < 3) {
            console.error('Error: Please provide an input file path');
            process.exit(1);
        }
        
        try {
            const input = await this.getEncodedInputStreamByPythonComment(filePath);
            const lexer = new PythonLexer(input);
            const tokens = new CommonTokenStream(lexer);
            const parser = new PythonParser(tokens);
        
            tokens.fill(); // Test the lexer grammar
            for (const token of tokens.tokens) {
                console.log(this.getTokenMetaDataWithTokenName(token));
            }
        
            parser.file_input(); // Test the parser grammar
            process.exit(parser.syntaxErrorsCount);
        } catch (error) {
            if (error instanceof Error) {
                console.error(`Error:\n${error.message}`);
            } else {
                console.error('An unknown error occurred');
            }
            process.exit(1);
        }
    }
}

grun4py.main(process.argv[2]);